### PR TITLE
Added encoding to user name and password

### PR DESCRIPTION
**Pull request by taylorstine**
I have an @ symbol in my mongo password. I realize that I can specify a custom class for the DatabaseAdapter, but I believe the decoding of a username and password should be built in to the default implementation.

### CODE REVIEW

File: ExportAdapter.js [hunk 1]
```diff
@@ -34,8 +34,19 @@ ExportAdapter.prototype.connect = function() {
     return this.connectionPromise;
   }
 
+  var usernameStart = this.mongoURI.indexOf('://') + 3;
+  var lastAtIndex = this.mongoURI.lastIndexOf('@');
+  var encodedMongoURI = this.mongoURI;
+  var username = null;
+  var password = null;
```
Comment (gfosco) [inline ExportAdapter.js]: It does not look like `username` or `password` are used... and it looks like the database uri is being somewhat validated here, so how about we throw an error if it's not?

**Comment by taylorstine** [comment]
@gfosco sorry for that, I've removed old references.  I've added a regex validation for the mongo URI.  It won't capture all invalid URIs, but it will at least validate the monogURI to be in the form we expect, [as specified here](https://docs.mongodb.org/manual/reference/connection-string/).

**Comment by gfosco** [comment]
Thanks for adding the regxr link.. very cool. :+1:

**Comment by themichael123** [comment]
This causes MongoDB Replica Sets to no longer work.

**Comment by taylorstine** [comment]
@themichael123 can you provide an example URI?

**Comment by themichael123** [comment]
@taylorstine Have a look here: https://docs.mongodb.org/manual/reference/connection-string/

I wouldn't use regex for this task.

**Comment by adamontherun** [comment]
@taylorstine @gfosco here's an example of a paid MongoLabs URI that isn't working: mongodb://test:testpass@ds056315-a0.mongolab.com:59325,ds059315-a1.mongolab.com:59315/testDBname?replicaSet=rs-ds059415

here's an example of a free sandbox MongoLabs URI that is working:
mongodb://test:testpass@ds053145.mongolab.com:59145/mytestserver

**Comment by adamontherun** [comment]
I confirmed that the MongoLabs replica set URI worked fine in Parse Server 2.0.4, the breaking change was in 2.0.5

**Comment by gfosco** [comment]
Let's get a fix PR sent in to parse-server to address this... @taylorstine are you interested in doing it?

**Comment by taylorstine** [comment]
@gfosco sure, but I could use a second opinion on the regex.  My aptitude for regex has kind of reached it's limit, maybe it should just be removed?

**Comment by adamontherun** [comment]
@gfosco @taylorstine I'd vote for removing the regex until we can be very confident in it. this bug breaks most parse server deployments that aren't running in the sandbox.

**Comment by drew-gross** [comment]
The approach we used on hosted parse was to remove one character at a time from the end of the connection string until it looked like a regular URI, and then we could use a regular URI parsing library on it, modify the contents as needed, and then rebuild the connection string that includes all the replica sets. That approach might work here as well. Make sure to add tests if you use that approach though.

**Comment by flovilmart** [comment]
We could also split the host of the URI around "@" char, and leave the last element being the host, and reconcat the n-1 first elements, then split on ":".

**Comment by jpgupta** [comment]
In my case it's moving from a sandbox to a production environment on MongoLabs that breaks it. 

It appears to be that the additional ?replicaSet argument added onto the URI fails the regex expression, I'd suggest taking it out as it's a little over-engineered. Testing the Mongo connection success/failure is probably worth while but it seems better to just trust people that they've entered a valid uri and will preserve future changes to the URI structure / arguments

**Comment by taylorstine** [comment]
https://github.com/ParsePlatform/parse-server/pull/297 I'm not confident in this approach, I don't want to break anything just because my password has a `@` sign

**Comment by flovilmart** [comment]
So could we split the full host on "@", then take the last part as the 'real' host and join the [0, n-1] parts together? Then from that, we split on ":" to get the username / password.

**Pull request merged**